addon/components/-private/base-pop-menu.js

```diff
@@ -0,0 +1,219 @@
+      this._triggerElement.removeEventListener('mouseleave', this._closePopoverHandler);
+    }
+
+    this._triggerElement.removeAttribute('data-popover-trigger');
+
+    document.body.removeEventListener('mouseup', this._handleBodyClick);
+  }
+
+  /**
+   * Inserts the popover and sets up the body click listener
+   */
+  _insertPopover() {
+    // Schedule these separately so that the element gets fully attached by setting
+    // `isOpen` to true, and only _then_ setting `isShowing` to true, triggering the
+    // CSS transition
+    run(() => this.set('isOpen', true));
+
+    document.body.addEventListener('mouseup', this._handleBodyClick);
+  }
+
+  /**
+   * Removes the popover and tears down the body click listener
+   */
+  _removePopover() {
+    run(() => this.set('isOpen', false));
+
+    document.body.removeEventListener('mouseup', this._handleBodyClick);
+  }
+
+  /**
+   * Triggers when the popover has entered the DOM and the API has been established,
+   * allowing us to add data attributes and event handlers and mark the trigger as active
+   *
+   * @param {PopperAPI} api - The API received from the underlying popper
+   */
+  @action
+  popoverOpened({ popperElement }) {
+    this._popperElement = popperElement;
+    this._popperElement.setAttribute('data-popover-content', guidFor(this));
+    this._triggerElement.classList.add('is-active');
+
+    if (this.get('triggerEvent') === 'hover') {
+      this._popperElement.addEventListener('mouseenter', this._openPopoverHandler);
+      this._popperElement.addEventListener('mouseleave', this._closePopoverHandler);
+    }
+  }
+
+  /**
+   * Triggers when the popover has been fully removed from the DOM (e.g. after animations)
+   * allowing us to teardown event handlers and mark the trigger element as inactive
+   */
+  @action
+  popoverClosed() {
+    if (this.get('triggerEvent') === 'hover') {
+      this._popperElement.removeEventListener('mouseenter', this._openPopoverHandler);
+      this._popperElement.removeEventListener('mouseleave', this._closePopoverHandler);
+    }
+
+    this._triggerElement.classList.remove('is-active');
+    this._popperElement.removeAttribute('data-popover-content');
+    this._popperElement = null;
+  }
+
+  // ----- Event Handlers -----
+
+  /**
+   * Opens the popover
+   */
+  _openPopoverHandler = () => {
+    if (!this.get('isOpen')) {
+      this._isOpening = true;
+      this._insertPopover();
+    }
+  }
+
+  /**
+   * Closes the popover
+   */
+  _closePopoverHandler = () => {
+    if (this.get('isOpen')) {
+      this._isOpening = false;
+      this._removePopover();
+    }
+  }
+
+  /**
+   * Handles a click on the body in general when a popover is open. If the clicked element is not
+   * a child of the popover, or is marked with `data-close`, closes the popover.
+   */
+  _handleBodyClick = ({ target }) => {
+    if (this._isOpening) {
+      this._isOpening = false;
+    } else if (closest(target, '.ice-base-pop-menu') === null) {
```

Comment:
I am still not satisfied with this approach though this is obviously an Ember bug. Unfortunately, we have to fix that. 

@LydiaBarnes can you use this branch and test the delete bug we saw last time and see if it happens?

Reply:
It may not even be an Ember bug though, it could be a bug in our app code. For instance, a very easy way to trigger this bug would be:

```js
button.addEventListener('click', () => {
  Ember.run(() => {
    // delete the item
    this.items.removeObject(item);
  });
});
```

By interfering with the normal runloop and forcing Ember to render eagerly, it would remove items before they should be removed, messing with the event flow.

Reply:
actually I'm 100% sure we're messing with the runloop in some way, see [this blogpost](https://medium.com/square-corner-blog/deep-dive-on-ember-events-cf684fd3b808). Native DOM events will _always_ fire before _any_ Ember actions are triggered, which is why the `body` handler should always trigger before Ember has had a chance to manipulate the DOM in any way.

If this happens in our app, we need to figure out why and fix it there.